### ISSUE TITLE
HIGH [$750]: Clean up the payment options on Pay button in New Dot

### ISSUE THREAD

**Issue by MitchExpensify** [Expensify/App#36301]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** v1.4.39-7
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**Expensify/Expensify Issue URL:** https://github.com/Expensify/App/issues/33967
**Issue reported by:** @anmurali 
**Slack conversation:** NA - 

## Action Performed:
1. Submit a request as User A to User B
2. As User B click "Pay with Expensify" on the request (Do nothing)
3. As User B click the “⌄” down chevron beside "Pay with Expensify" (Do nothing) 

## Expected Result:

The same options should be presented when clicking "Pay with Expensify" or the “⌄” down chevron beside it. 

**This is the new design we want to implement lifted from https://github.com/Expensify/App/issues/33967:** 

> Pull all the options into one single dropdown labeled `Pay <currency><amount>`, which when pressed opens for the user to choose:
> 
> `Pay with business bank account`
> `Pay with personal bank account`
> `Pay with debit card`
> `Pay elsewhere`
> 
> - Let's remember payment preference based on the first payment by request type. E.g. IOUs vs Expenses. 
> 
> Said another way, if you paid an IOU using a debit card, then next time you pay an IOU, unless you pressed on the down caret (dropdown), we assume you mean the same payment device and process the payment with the debit card. But if you then pay an expense in a workspace, we don't assume but ask you to choose a payment method. 
> 
> - When defaulting payment method for expenses, use the workspace default if one is set in workspace settings.
> 
> If there is a default at the workspace level, and the admin has access to that, use that.
> 
> - When defaulting payment method for expenses, make it workspace specific. 
> 
> Said another way, if you pay an expense on workspace A, we default you to that payment device the next time (unless you press the dropdown). But then if you go to pay an expense on workspace B for the first time, we do not assume.

## Actual Result:

Two different sets of options are presented to the user:

A:
![image](https://github.com/Expensify/App/assets/36425901/af200683-44bc-493e-9851-e5d4de814388)

B:
![image](https://github.com/Expensify/App/assets/36425901/33b6eeaa-e498-487f-9663-635c374578a7)

## Workaround:
None

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [x] Android: mWeb Chrome
- [x] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~016cb0dade2c28979f</li>
        <li>Upwork Job ID: 1758260187340271616</li>
        <li>Last Price Increase: 2024-06-11</li>
    </ul>
</details>

**Comment by MitchExpensify** [comment]
Slack convo https://expensify.slack.com/archives/C05NJ4SLBMF/p1707362622166499

**Comment by joekaufmanexpensify** [comment]
[Discussing ](https://expensify.slack.com/archives/C05NJ4SLBMF/p1707858472345099?thread_ts=1707362622.166499&cid=C05NJ4SLBMF)

**Comment by MitchExpensify** [comment]
Let's run with this issue!

**Comment by MitchExpensify** [comment]
Let's hunt down proposals @joekaufmanexpensify 🦾

**Comment by ikevin127** [comment]
> [!NOTE]
> Considering the amount of work involved - I'm willing to work on this for no less than $1000.

## Proposal

### Please re-state the problem that we are trying to solve in this issue.

The `Pay` button in ND has two sets of options:
- pressing the button brings up payment options
- pressing on the drop down brings up options to pay elsewhere

This UX is pretty confusing.

### What is the root cause of that problem?

New feature / refactoring / clean-up.

### What changes do you think we should make in order to solve the problem?

From https://github.com/Expensify/App/issues/33967: follow the specs / conversation carefully and build on @tgolen's draft PR https://github.com/Expensify/App/pull/34689 to accomplish the **Expected result** as stated.

Next steps in moving forward using the draft PR in order to get this shipped -> from https://github.com/Expensify/App/issues/33967#issuecomment-1915035254:

> There are multiple sets of options as mentioned in the original description. One set of options is in front of the KYCPaywall (pay with expensify, pay elsewhere), and one set of options is behind the KYCPaywall (bank account, personal bank account, debit card). In order to have all options in a single dropdown, then the way the dropdown interacts with the KYCPaywall needs to be refactored.

**Comment by brandonhenry** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The problem is the inconsistency and clutter in the payment options presented in New Dot when a user attempts to pay a request. The current implementation separates payment methods into different actions, which can confuse users. The goal is to streamline the payment process into a single, cohesive flow.

### What is the root cause of that problem?
The root cause is the current design, which does not unify payment options under a single action. There's a lack of intuitive UI flow that remembers the user's previous payment preferences, leading to a repetitive selection process for each transaction.

### What changes do you think we should make in order to solve the problem?

- Redesign the payment button to a single dropdown labeled "Pay <currency><amount>", which will consolidate all payment options into one list.
- Implement a feature to remember the user's payment preference based on the first payment by request type (IOUs vs. Expenses) and default to that method for subsequent payments unless the user specifies otherwise.
- For expenses, default the payment method to be workspace-specific, using the workspace default settings if set and the user has access to it.
- Ensure that this preference is set at the workspace level and does not cross-apply to other workspaces to maintain individual workspace preferences.

**Comment by HarlemSquirrel** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

There are two sets of options presented in the payment button dropdown shown to users who receive a request. This can cause confusion so we would like to consolidate the options. Clicking either the button text or the chevron should show all available options.

### What is the root cause of that problem?

The button options are set on `AddPaymentMethodMenu` has two items in the `buttonOptions` array. The second sets of options which includes "elsewhere" is only accessible by clicking on the chevron. 

### What changes do you think we should make in order to solve the problem?


We should remove the second item in `buttonOptions` and add the elsewhere option to `AddPaymentMethodMenu`.

**Comment by HarlemSquirrel** [comment]
Contributor details
Your Expensify account email: kevin@mccormack.tech
Upwork Profile Link: https://www.upwork.com/freelancers/~01435a01fa4343ac93

**Comment by joekaufmanexpensify** [comment]
Proposals pending review

**Comment by tgolen** [comment]
@all for your proposals, please be sure you have looked at the code and understand how the `KYCPaywall` is used and what refactorings are necessary to have all options be in a single dropdown. You can see my comments in https://github.com/Expensify/App/issues/33967#issuecomment-1915035254 for a little more context.

**Comment by tgolen** [comment]
@ikevin127 it sounds like you are on the right track! Do you have any idea of what needs to be done with the `KYCPaywall`?

**Comment by HarlemSquirrel** [comment]
Are we talking about `KYCWall`? 

Looks like we no longer want to save the preferred payment method so we can change it to do the same thing as when the button is pressed to call `selectPaymentType()` or remove the options dropdown by swapping `ButtonWithDropdownMenu` for `Button`

**Comment by ikevin127** [comment]
@tgolen Won't pretend to fully understand the system but from a quick look at the code I can tell the following:

1. We're using the `AddPaymentMethodMenu` component within the `KYCWall` which in the case where a user goes with the **Pay with Expensify** option (selected from dropdown) -> the dropdown options will be: Personal / Business BA (if the user can use any of the 2) and Card. When one of these are chosen -> user needs to pass the KYCWall (if didn't already).

https://github.com/Expensify/App/blob/bea6aafa5e1b4a80462c3da1f3c56325d3a3977d/src/components/AddPaymentMethodMenu.js#L68

2. Within the `SettlementButton` we use the `KYCWall` which exposes the following dropdown options:
- **Pay with Expensify**:  if chosen and we click on the button (not dropdown) -> the Personal / Business BA (if the user can use any of the 2) and Card options are presented which require passing the KYCWall (if didn't already)
- **Pay elsewhere**: if chosen from the dropdown and we click on the button (not dropdown) -> KYCWall is not involved and the request will be marked as paid right away.

This being said, if we're looking at the following **Expected result**:

> Pull all the options into one single dropdown labeled Pay <currency><amount>, which when pressed opens for the user to choose:
> `Pay with business bank account`
> `Pay with personal bank account`
> `Pay with debit card`
> `Pay elsewhere`

then we would need to pass all the options and adjust the current logic of the `KYCWall` from within the `SettlementButton` to display all of the above options when clicked on the dropdown arrow and selecting any -> the button (left side) will call the action of the selected dropdown option as we currently do for both non KYC / KYCWall options.

This can be a pretty congnitively complex endeavour since it would involve carefully considering cases like:
- remembering the payment preference based on the first payment by request type (IOUs vs Expenses)
- when defaulting payment method for expenses, use the workspace default if one is set in workspace settings
- when defaulting payment method for expenses, make it workspace specific
- if user has not paid using any of the options before (first time paying) and clicks on the left side of the button -> show the dropdown

but an intresting one to take on nonetheless!

**Comment by brandonhenry** [comment]
@tgolen makes sense! here is my updated proposal, with more details on what I think needs to change in order to meet specs:


# Updated Technical Proposal

## Problem Statement
The 'Pay' button currently presents disjointed payment options, which we want to consolidate into a single dropdown for a smoother user experience.

## Examination of the Current Code
The `KYCWall` component acts as a gatekeeper for payment actions, ensuring that users meet KYC requirements. The `AddPaymentMethodMenu` component presents payment options to the user.

1. **KYCWall Component (KYCWall.js):**
   - Currently handles the display logic of payment methods and KYC checks.
   - Interacts with the `AddPaymentMethodMenu` to present payment options.
   - Source: [KYCWall.js](https://github.com/Expensify/App/blob/main/src/components/KYCWall/BaseKYCWall.tsx#L6)

2. **AddPaymentMethodMenu Component (AddPaymentMethodMenu.js):**
   - Renders the popover menu with payment options.
   - Needs to be refactored to include a unified list of payment methods.
   - Source: [AddPaymentMethodMenu.js](https://github.com/Expensify/App/blob/main/src/components/AddPaymentMethodMenu.js)

## Proposed Changes
I propose refactoring these components to create a unified dropdown while ensuring compliance with KYC requirements.

1. **Integrate Payment Options into a Single Dropdown:**
   - Refactor the `menuItems` prop within `AddPaymentMethodMenu` to include all payment methods post-KYC check.
   - Implement dynamic rendering based on KYC status and user preferences.
     ```javascript
     menuItems={[
       ...PaymentUtils.getPaymentMethods(bankAccountList, fundList, shouldShowPersonalBankAccountOption, shouldIncludeDebitCard).map(paymentMethod => ({
         text: translate(paymentMethod.textKey),
         icon: paymentMethod.icon,
         onSelected: () => selectPaymentMethod(paymentMethod.type),
       })),
     ]}
     ```

2. **Enhance State Management:**
   - Use the Onyx library to manage and remember user's payment preferences.
   - Reference [ONYXKEYS](https://github.com/Expensify/App/blob/main/src/ONYXKEYS.js) to set and retrieve user preferences.
     ```javascript
     Onyx.set(ONYXKEYS.USER_PAYMENT_PREFERENCES, { lastUsedMethod: selectedMethodType });
     ```

3. **Address First-Time Payment Flow:**
   - For first-time payments, trigger the dropdown and prompt the user to select a payment method.
   - Incorporate a check for first-time users:
     ```javascript
     if (PaymentUtils.isFirstTimePaying(userWallet)) {
       showPaymentMethodSelection();
     }
     ```

4. **Workspace-Specific Defaults:**
   - Modify the logic to set and retrieve workspace-specific defaults.
   - Integrate with backend services for workspace preferences:
     ```javascript
     API.getWorkspacePreferences(policyID).then((preferences) => {
       setWorkspacePaymentDefaults(preferences.defaultPaymentMethod);
     });
     ```

5. **Refactor KYCPaywall Interaction:**
   - Alter how the `KYCPaywall` interacts with the dropdown, especially considering the unified payment options.
   - Consider moving the KYCPaywall check to the `onItemSelected` event to ensure seamless user experience.

## Additional Considerations

The `KYCWall` component is crucial because it determines whether the user should be shown the `AddPaymentMethodMenu` or be taken through the KYC flow. The key to refactoring this interaction is to ensure that the KYCPaywall's logic seamlessly integrates with the dropdown's display logic, enabling or disabling payment options as necessary.

### Refactor KYCPaywall Interaction:

1. **KYCPaywall and Dropdown Integration:**
   - Enhance the `selectPaymentMethod` callback within `KYCWall` to handle KYC validations before setting the selected payment method. This ensures that the KYC flow is completed before a payment method is considered valid.
     ```javascript
     const selectPaymentMethod = useCallback((paymentMethod: PaymentMethod) => {
       if (PaymentUtils.needsKYC(paymentMethod)) {
         Wallet.startKYCFlow(paymentMethod);
         return;
       }
       onSelectPaymentMethod(paymentMethod);
       // ... existing logic ...
     }, [onSelectPaymentMethod]);
     ```

2. **Conditional Rendering in `AddPaymentMethodMenu`:**
   - Incorporate KYC status checks directly in the rendering logic of the `AddPaymentMethodMenu`. This might involve using a new prop to indicate which payment methods are available post-KYC and conditionally rendering them.
     ```javascript
     menuItems={PaymentMethods.getAvailableMethods(userWallet, reimbursementAccount).map((method) => ({
       text: translate(method.name),
       icon: method.icon,
       onSelected: () => {
         if (method.requiresKYC && !userWallet.hasPassedKYC) {
           Wallet.promptForKYC();
           return;
         }
         onItemSelected(method.type);
       },
     }))}
     ```

3. **User Feedback for KYC-Required Methods:**
   - In the `AddPaymentMethodMenu`, provide feedback for payment methods that require KYC. For instance, this could be a tooltip or a disabled state with an explanation that appears when the user tries to select an option that requires KYC completion.
     ```javascript
     menuItems={...}
     renderItem={(item) => (
       <MenuItem
         icon={item.icon}
         text={item.text}
         isDisabled={item.requiresKYC && !userWallet.hasPassedKYC}
         onPress={() => {
           if (item.requiresKYC && !userWallet.hasPassedKYC) {
             showAlert('Please complete KYC to use this payment method.');
             return;
           }
           item.onSelected();
         }}
       />
     )}
     ```

By integrating these changes, we ensure that the KYCPaywall remains an integral part of the payment process without disrupting the user experience of having all payment options in one place. 

This is just a rough idea that I was able to gather from my few hours of going through these components, but I believe this is on the right track for what is being requested. Happen to dive in further if needed!

**Comment by allroundexperts** [comment]
Looking into the proposals today 👀

**Comment by joekaufmanexpensify** [comment]
Great. TY!

**Comment by allroundexperts** [comment]
Thanks for your proposals everyone. While @ikevin127 seems to be moving on in the right direction, I feel like his proposal is too abstract. 

@brandonhenry's [proposal](https://github.com/Expensify/App/issues/36301#issuecomment-1949090025) on the other hand has a lot more detail on how he'll be achieving the unification of the payment options. As such, let's go with them.

🎀 👀 🎀 C+ reviewed

**Comment by brandonhenry** [comment]
Thanks! I should have this PR ready by EOW. Will update if anything comes up

**Comment by joekaufmanexpensify** [comment]
Great. TY!

**Comment by joekaufmanexpensify** [comment]
Generally, we want PRs to be very thorough. But curious what @allroundexperts and @bondydaa think about this [specific question ](https://github.com/Expensify/App/issues/36301#issuecomment-1962427325), as I'm not exactly sure of the implications of the two options here.

**Comment by bondydaa** [comment]
I'm not sure I understand the question (or the system enough maybe 😅). feel free to tag @allroundexperts in a comment on your PR if you did something but aren't sure about it and we can discuss it there.

**Comment by joekaufmanexpensify** [comment]
@brandonhenry @allroundexperts is there an ETA for this to get to full review? I see the PR is still a WIP, and there are some outstanding Q's on it.

**Comment by allroundexperts** [comment]
> @joekaufmanexpensify so I was doing some digging on this and had a question since this is my first ticket as a contributor. I was wondering how thorough my PR needs to be? I was digging around, and noticed that is a bit of a confusing matrix between the two types of payments options
> 
> Specifically, I need to update this line
> 
> https://github.com/Expensify/App/blob/e9fb49d9f41972a04bf07905b228314a3fdce8d4/src/types/onyx/OriginalMessage.ts#L5
> 
> to include `CONST.PAYMENT_METHODS` in order for me to consolidate the dropdowns.
> 
> https://github.com/Expensify/App/blob/9aa0509b3b155e9adbb95d8a32f78f7d3822c350/src/CONST.js#L435
> 
> That said, I have the UI working exactly like it works right now, but the visuals match what is requested in this ticket. Should I stop there since functionality is the same or should I keep digging / remove / rework how the system is setup to handle the different payment options (which now that i write this out, sounds like it may have some backend implications)

Generally, I think over-engineered work brings in problems down the road. If your work is having backend implications, its best to mention that when writing the proposal. For this case though, can you elaborate on why we'll need the backend changes?

**Comment by joekaufmanexpensify** [comment]
Looks like PR is pending review from C+

**Comment by mallenexpensify** [comment]
@brandonhenry per [CONTRIBUTING.md,](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md?rgh-link-date=2024-04-01T17%3A24%3A34Z#:~:text=New%20contributors%20are%20limited%20to%20working%20on%20one%20job%20at%20a%20time%2C%20however%20experienced%20contributors%20may%20work%20on%20numerous%20jobs%20simultaneously.)

> New contributors are limited to working on one job at a time, however experienced contributors may work on numerous jobs simultaneously.

You currently have [6 issues open](https://github.com/Expensify/App/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+assignee%3Abrandonhenry+) with non merged or closed. Please do not submit proposals for new issues until **all** six PRs have been merged.

Also [from CONTRIBUTING.md](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md?rgh-link-date=2024-04-01T17%3A24%3A34Z#:~:text=Daily%20updates%20on,original%20contract%20terminated.)

> Daily updates on weekdays are highly recommended. If you know you won’t be able to provide updates within 48 hours, please comment on the PR or issue stating how long you plan to be out so that we may plan accordingly. We understand everyone needs a little vacation here and there. Any issue that doesn't receive an update for 5 days (including weekend days) may be considered abandoned and the original contract terminated.

Please keep the issue and PR moving along with urgency, along with all others.  You were assigned here on February 21 and it's now April 1.

**Comment by joekaufmanexpensify** [comment]
PR still in review

**Comment by anmurali** [comment]
Can someone share a video of how this works post-fix? From the dev environment perhaps? We are actively discussing this design again in slack, and we want to make sure where we landed [there](https://expensify.enterprise.slack.com/archives/CSL3XBCCR/p1713362838125699?thread_ts=1713358387.605059&channel=CSL3XBCCR&message_ts=1713362838.125699) is how we fixed it here.

**Comment by dannymcclain** [comment]
@shawnborton Yeah I think so.

The only thing I would add is: do we want to clarify the specific payment method used last vs just `Pay with Expensify`?
So perhaps something like `Pay with Bortibank (1234)` or `Pay with Visa (5678)`? Or would tapping `Pay with Expensify` just bring up any of those options you already have stored?

**Comment by shawnborton** [comment]
> My current implementation combines the two menus, checking if you have added a payment option, and displays Add X Payment option if you haven't (but all are always shown as in https://github.com/Expensify/App/issues/36301#issuecomment-2061625926)

My thinking is that if you have never added a payment option, we could just use a generic word "Pay" and tapping that part of the button, or the arrow, would launch the same menu of all options. Does that make sense and sound like a good idea?

**Comment by shawnborton** [comment]
Sounds great! Also want to get a gut check from @anmurali and @puneetlath and @Expensify/design that they are on board with this approach too

**Comment by shawnborton** [comment]
Seems like we have a lot of thumbs, let's do it

**Comment by joekaufmanexpensify** [comment]
TY! I'll bump the message to other channels if it doesn't get any traction.

**Comment by joekaufmanexpensify** [comment]
@allroundexperts @brandonhenry are we close to getting this merged you think?

**Comment by allroundexperts** [comment]
@joekaufmanexpensify Most of my urgent issues are finished so I should be able to speed things up. Hoping to get this in by this week.

**Comment by joekaufmanexpensify** [comment]
Hey @brandonhenry , we were [discussing this one](https://expensify.slack.com/archives/CSL3XBCCR/p1715784975621439?thread_ts=1713358387.605059&cid=CSL3XBCCR) internally a bit, and want to make sure our understanding of all of the cases is correct. Could you please outline screenshots/video showing someone paying (both the big part of the pay button, and the arrow options), for the following cases in the issue:

1. No payment method. 
2. Has payment method but has never paid before
3. Has payment method and has paid before.

LMK if you have any Q's. TY!

**Comment by joekaufmanexpensify** [comment]
TY!

**Comment by joekaufmanexpensify** [comment]
@brandonhenry any update [on this](https://github.com/Expensify/App/issues/36301#issuecomment-2113006581)? I'd like to pass this to the rest of the team as soon as I can.

**Comment by davidcardoza** [comment]
@brandonhenry Can you please provide an update here?

**Comment by joekaufmanexpensify** [comment]
Sorry to hear that, and hope you're okay! Good luck with everything!

**Comment by davidcardoza** [comment]
Wow, hope you and your family are doing okay.

**Comment by joekaufmanexpensify** [comment]
Hey @brandonhenry, checking in. Hope you're doing okay! Are you still working on settling everything related to the natural disaster in Houston? No worries at all, if so, just wanted to see where you're at.

**Comment by joekaufmanexpensify** [comment]
Course, and thanks for confirming!

**Comment by davidcardoza** [comment]
@brandonhenry any update on this one?

**Comment by allroundexperts** [comment]
@brandonhenry That seems more of a product question. @joekaufmanexpensify Can we clarify the expected behaviour after #41859 got merged?

**Comment by joekaufmanexpensify** [comment]
[Asked internally ](https://expensify.slack.com/archives/CSL3XBCCR/p1717521700980619?thread_ts=1717198394.028729&cid=CSL3XBCCR)

**Comment by joekaufmanexpensify** [comment]
>Also, as a side note, I think that there has been a lot of new additions and discussion points on this ticket. I'd like to request an increase for the payment just based on complexity and subsequent amount of testing to ensure this works

Sure, I can look into this and report back.

**Comment by joekaufmanexpensify** [comment]
> unfortunately, I am still not able to verify functionality here. i even tried to add my real debit card and verify my real info, i am still not able to enable my wallet to test this...

Is this impacting you testing only certain parts of the flow, or basically everything?

**Comment by joekaufmanexpensify** [comment]
Hmmm, got it. If it's only a certain part of the flow, maybe it would be easiest if the internal assigned engineer (@bondydaa) just makes sure that this piece works while reviewing? LMK if you're cool with that Bondy. 

I feel like it could be more work to bypass the wallet verification process.

**Comment by joekaufmanexpensify** [comment]
>Also, as a side note, I think that there has been a lot of new additions and discussion points on this ticket. I'd like to request an increase for the payment just based on complexity and subsequent amount of testing to ensure this works

[Chatted](https://expensify.slack.com/archives/C01SKUP7QR0/p1718121502826089?thread_ts=1718120820.314419&cid=C01SKUP7QR0) about this internally, and we landed on bumping the price here to $750!

**Comment by joekaufmanexpensify** [comment]
As an FYI, I am about to go OOO until 2024-06-25. I don't think payment will be due before then, so not adding an additional BZ assignee. However, if you need anything before then, @davidcardoza or @bondydaa should be able to help.

**Comment by davidcardoza** [comment]
@brandonhenry Can you please provide ab update when you have a minute?

**Comment by joekaufmanexpensify** [comment]
@brandonhenry before we finish the PR, is it possible still to send [these screenshots/videos](https://github.com/Expensify/App/issues/36301#issuecomment-2113006581) so I can share them with the rest of the team? TY!

**Comment by davidcardoza** [comment]
I am not certain that the need to process a payment is necessary to complete this issue. The pay button experience is used pre-purchase.

**Comment by shawnborton** [comment]
Are we able to align the popover to the right-edge of the button below it?
![CleanShot 2024-06-25 at 18 02 07@2x](https://github.com/Expensify/App/assets/2319350/905e3896-fd4d-4fcb-919d-8d1e5a786731)

**Comment by joekaufmanexpensify** [comment]
PR in review and seems close to merge!

**Comment by joekaufmanexpensify** [comment]
Thanks for the update! Glad to hear you made it through the storm okay.

**Comment by joekaufmanexpensify** [comment]
Sounds good. TY!

**Comment by joekaufmanexpensify** [comment]
@brandonhenry do you think this one will be merged soon?

**Comment by joekaufmanexpensify** [comment]
TY!

**Comment by joekaufmanexpensify** [comment]
TY!

**Comment by joekaufmanexpensify** [comment]
[Discussing ](https://expensify.slack.com/archives/C01GTK53T8Q/p1723033062800229)

**Comment by joekaufmanexpensify** [comment]
[PR](https://github.com/Expensify/App/pull/37174) still in review. @brandonhenry is there an ETA for when this will be ready for re-review?

**Comment by joekaufmanexpensify** [comment]
PR is merged!

**Comment by bondydaa** [comment]
this was reverted on https://github.com/Expensify/App/pull/48117 b/c it caused https://github.com/Expensify/App/issues/48103

**Comment by danielrvidal** [comment]
@brandonhenry can you comment on the issue it caused so we can get this fixed and then merged once more?

**Comment by bondydaa** [comment]
yeah I'm a bit confused by that one too 🤷 are you able to reproduce the gradient locally?

we can probably revert the revert https://github.com/Expensify/App/pull/48117 so get your changes back in.

or if it's easier you can also just open up a new PR.

**Comment by bondydaa** [comment]
left a note on that other issue here https://github.com/Expensify/App/issues/48103#issuecomment-2318376388.

you should be able to easily open up a revert PR from this PR https://github.com/Expensify/App/pull/48117, if not let me know and I can give it a go.

**Comment by joekaufmanexpensify** [comment]
I think next step here is still to revert the revert and get the PR back out. Is that right @brandonhenry ?

**Comment by bondydaa** [comment]
oh weird 🤷 you don't see this button? 
<img width="2108" alt="image" src="https://github.com/user-attachments/assets/f99c403a-5f26-4c26-b056-ae3a6d359059">

anywho, i created it here https://github.com/Expensify/App/pull/48421

**Comment by allroundexperts** [comment]
This is a straight revert, so I don't think a review is necessary again here.

**Comment by joekaufmanexpensify** [comment]
Sweet. TY!

**Comment by joekaufmanexpensify** [comment]
Yep, agreed.

**Comment by trjExpensify** [comment]
I'm trying to figure out why I only see the `Pay with business bank account` option in the button now, and I think it's maybe because of the work here? I should retain the option to be able to "Pay elsewhere" as well to mark it as manually reimbursed. CC: @Expensify/design @JmillsExpensify 

<img width="1134" alt="image" src="https://github.com/user-attachments/assets/32e25433-92a2-4f06-80b5-07e998af33f9">

<img width="1142" alt="image" src="https://github.com/user-attachments/assets/1fb44a59-624a-4c3a-beab-fc3bf8f1d5f6">

**Comment by shawnborton** [comment]
Hmm yeah, there should still be the dropdown arrow portion of the split button to choose a different method if you want to.

**Comment by JmillsExpensify** [comment]
Agree with that.

**Comment by bondydaa** [comment]
@brandonhenry @allroundexperts are you able to take a look at those reports above?

And we also have this one that popped up https://github.com/Expensify/App/issues/48560 which I'm not sure if related or not... ideally I don't think we want/need to revert right now but if https://github.com/Expensify/App/pull/37174 is the root cause maybe we do 🤷

**Comment by trjExpensify** [comment]
FWIW, I think the Test/QA steps in https://github.com/Expensify/App/pull/37174 are super light given the changes here that span a lot further than just pay options for iouReports. I'm not surprised this keeps getting reverted. 

Revisiting the OP to try and max sense of this one. 

>Pull all the options into one single dropdown labeled Pay <currency><amount>, which when pressed opens for the user to choose:
>
>Pay with business bank account
>Pay with personal bank account
>Pay with debit card
>Pay elsewhere

I _assume_ those are the pay options for iouReports only, but it still suggests the button should be a button displayed with a menu for `Pay $amount v`. 

Then in the OP, there’s some additional logic changes for the stickiness..

> - Let’s remember payment preference based on the first payment by request type. E.g. IOUs vs Expenses.
>Said another way, if you paid an IOU using a debit card, then next time you pay an IOU, unless you pressed on the down caret (dropdown), we assume you mean the same payment device and process the payment with the debit card. But if you then pay an expense in a workspace, we don’t assume but ask you to choose a payment method.

This makes sense, as some pay button options on iouReports aren’t applicable or available on expenseReports on workspace expenses (i.e PBA or Debit Card). Your preferred payment method for each of those report types can conceivably be different. 

Did this get tested and implemented in the PR, or is it outstanding? 

>But if you then pay an expense in a workspace, we don’t assume but ask you to choose a payment method

I think this last line in that bullet doesn't make sense though. It also contradicts the next bullet down, so let's move on to that bullet. 

> - When defaulting payment method for expenses, use the workspace default if one is set in workspace settings.
>If there is a default at the workspace level, and the admin has access to that, use that.

This sounds like something we are doing already by using the VBBA set in the workspace settings when you’re the reimburser viewing an expenseReport that needs reimbursing. You can choose to `Pay elsewhere` to manually mark it as reimbursed, but you can’t choose a different bank account to pay the expense report?

> - When defaulting payment method for expenses, make it workspace specific.
> Said another way, if you pay an expense on workspace A, we default you to that payment device the next time (unless you press the dropdown). But then if you go to pay an expense on workspace B for the first time, we do not assume.

Same here, doesn't this come out of the box with setting a VBBA to reimburse from in the workspace settings?

Overall, it seems like this has got a bit wonky in implementation perhaps. I'm leaning towards reverting and revisiting this with a clearer picture of the scope of implementation.

**Comment by JmillsExpensify** [comment]
Agree with Tom

**Comment by trjExpensify** [comment]
I just came across another issue, we're failing to pay any expense reports in product. Investigation [here](https://github.com/Expensify/App/issues/48545#issuecomment-2331395460).

@mountiny can you help us revert [this PR](https://github.com/Expensify/App/pull/37174) and CP it?

**Comment by trjExpensify** [comment]
>do you have thoughts on how to proceed? 

1. Revert the necessary PR/PRs so we unbreak the product
2. Take the high level design back to Slack and crystalise mockups for every flow and instance of the pay button and payment method options for: 
     - iouReports 
     - expenseReports
     - invoiceReports
     - Pay button behaviour inside each of those report types
     - Pay button behaviour on the REPORTPREVIEW of each of those report types
     - Updates to lastPaymentMethod "sticky" logic for each report type
3. Agree on the detailed implementation and rollout plan to accommodate where all the changes need to be made on the FE and BE. 
4. Write manual tests for each scenario, to add those to the PR to be tested and assist during the development to check the intended behaviour
5. Update this issue OP to reflect the finalised scope of work. 
6. Code it, review it, ship it.

**Comment by joekaufmanexpensify** [comment]
Maybe a quick design doc could even make sense? The amount of discussion on this issue about implementation, and the slack threads linked out to already further discussing kinda shows this isn't really a straightforward change. It might benefit from making sure we are holistically considering all angles of the solution.

**Comment by trjExpensify** [comment]
100%

**Comment by joekaufmanexpensify** [comment]
Going to bring next steps to slack very soon.

**Comment by joekaufmanexpensify** [comment]
Okay, we discussed this internally and here's where we landed. We're going to: 

- Pay out the fully bounty here to both the @brandonhenry and @allroundexperts. A lot of work went into this one, and in hindsight, this was a much more complex change than initially described in the OP of the issue. Any issues with the implementation would have likely been obviated had we designed this most holistically, rather than going straight to a GH issue. 
- Close the GH issue. 
- Take a fresh look at this internally, with a full design doc to make sure we handle all of the permutations of the various impacted flows before moving to implementation. 

LMK if anyone has questions on that. Otherwise will tackle the payment steps later today!

**Comment by joekaufmanexpensify** [comment]
Payment summary here: 

- @brandonhenry - $750 for PR via Upwork. 
- @allroundexperts - $750 for C+ via NewDot.

**Comment by joekaufmanexpensify** [comment]
@brandonhenry offer for $750 sent via Upwork

**Comment by joekaufmanexpensify** [comment]
@allroundexperts you're good to request payment here

**Comment by brandonhenry** [comment]
@joekaufmanexpensify thank you!!

**Comment by joekaufmanexpensify** [comment]
Upwork job closed.

**Comment by joekaufmanexpensify** [comment]
Closing this one out as all that's left is for @allroundexperts to request payment in Newdot. We'll be continuing discussion on the longer term solution for this in Slack!

**Comment by JmillsExpensify** [comment]
$750 approved for @allroundexperts
